### ISSUE TITLE
Discrepancy between logged custom_stoploss value and actual stop_loss exit in backtest results for a subset of trades

### ISSUE THREAD

**Issue by Sakuleta** [freqtrade/freqtrade#11802]
Describe your environment
Operating system: Windows 11
Python Version: Python 3.13.2
CCXT version: CCXT 4.4.77
Freqtrade Version: 2025.4 

Describe the problem:
When backtesting a strategy that utilizes a custom_stoploss function, a small subset of trades (18 out of 426 in the provided backtest) are reported in the backtest summary table as closing due to the stop_loss exit reason with an average profit percentage of approximately -101.52%. This loss percentage is consistent with the strategy's global fallback stoploss = -0.99 setting.

However, the custom_stoploss function in the strategy has been designed with detailed logging for all conditions that would lead to it returning this fallback value. The full backtest logs, generated with --cache none, do not show any of these explicit warning or error log messages from within the custom_stoploss function that would explain why these 18 specific trades defaulted to the -0.99 stop.

Instead, for trades where custom_stoploss logs are available (including problematic trades identified from the JSON results manually), the logs indicate that the function correctly calculates and clamps the stop-loss to a more conservative value (e.g., -0.35 of stake, or tighter if the calculated loss percentage based on ATR and leverage is already within the defined min_abs_stake_loss_cap and max_abs_stake_loss_cap).

This suggests a potential discrepancy where the valid stop-loss value returned by custom_stoploss might not be applied as expected for this specific subset of trades, or the fallback stop-loss is being triggered by a mechanism not logged by the custom function, despite use_custom_stoploss = True.

Steps to reproduce:
Use the strategy file AdvancedStrategy_4h.py https://gist.github.com/Sakuleta/e19acbc948c15c5fdce74f8dc3fc584d
Use the config.json file https://gist.github.com/Sakuleta/3e8ec3649b6769c8df0189185664e487
Run backtesting using the command: 
docker compose run --rm freqtrade backtesting --config user_data/config.json --strategy AdvancedStrategy_4h --timerange 20200527-20250526 --cache none

Observed Results:
Backtest Summary Table ("Exit Reason Stats"):

```
┏━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Enter Tag ┃        Exit Reason ┃ Trades ┃ Avg Profit % ┃ Tot Profit USDT ┃ Tot Profit % ┃ Avg Duration ┃  Win  Draw  Loss  Win% ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│           │                roi │    348 │         7.92 │        2753.932 │        137.7 │      0:03:00 │  348     0     0   100 │
│           │ trailing_stop_loss │     60 │       -11.19 │        -671.199 │       -33.56 │      0:36:00 │    2     0    58   3.3 │
│           │          stop_loss │     18 │      -101.52 │       -1827.339 │       -91.37 │      0:00:00 │    0     0    18     0 │
│     TOTAL │                    │    426 │         0.61 │         255.394 │        12.77 │      0:08:00 │  350     0    76  82.2 │
└───────────┴────────────────────┴────────┴──────────────┴─────────────────┴──────────────┴──────────────┴────────────────────────┘
```

This indicates 18 trades closed at approximately -100% loss on stake, consistent with the -0.99 fallback stoploss.

custom_stoploss Log Output (from the full log file generated with --cache none):
```
For the vast majority of trades, and specifically for problematic TradeIDs identified from the JSON results (e.g., TradeID 23 XRP/USDT, LTC/USDT, TradeID 30 XLM/USDT), the logs show custom_stoploss calculating appropriate clamped values:
2025-05-26 22:12:27,190 - AdvancedStrategy_4h - INFO - CustomStop for XLM/USDT:USDT (TradeID: 23): Leverage=25.5x, ATR=0.00156, OpenRate=0.07489, ATR*Multiplier_PriceDistance=0.00313, CalculatedAbsoluteStakeLoss%=106.48%, ClampedAbsoluteStakeLoss%=35.00%, FinalDynamicStopProfit%=-35.00%
2025-05-26 22:12:27,240 - AdvancedStrategy_4h - INFO - CustomStop for XRP/USDT:USDT (TradeID: 28): Leverage=22.5x, ATR=0.00392, OpenRate=0.25250, ATR*Multiplier_PriceDistance=0.00784, CalculatedAbsoluteStakeLoss%=69.86%, ClampedAbsoluteStakeLoss%=35.00%, FinalDynamicStopProfit%=-35.00%
```

[backtest-result-2025-05-26_22-12-33.json](https://github.com/user-attachments/files/20446383/backtest-result-2025-05-26_22-12-33.json)
[backtest-result-2025-05-26_22-12-33_config.json](https://github.com/user-attachments/files/20446382/backtest-result-2025-05-26_22-12-33_config.json)

Absence of Fallback Warning/Error Logs: The full log file does not show any of the expected logger.warning or logger.error messages from within the custom_stoploss function (e.g., "ATR is zero or negative", "Leverage is zero or None", etc.) that would correspond to these 18 specific trades and explain why the fallback self.stoploss = -0.99 was returned by the function.

Trade Data from Results JSON:
Trades listed with "exit_reason": "stop_loss" in the JSON results file consistently show a "profit_ratio" around -1.0 (e.g., -1.0152, -1.0389) and a "stop_loss_ratio" of -0.99. Example for one such trade:
```
{
    "pair": "LTC/USDT:USDT",
    "stake_amount": 99.99923333,
    "open_date": "2020-10-06 16:00:00+00:00",
    "open_rate": 47.9,
    "close_rate": 46.32,
    "profit_ratio": -1.021555817834983, 
    "exit_reason": "stop_loss", 
    "stop_loss_ratio": -0.99,
    "trade_id": 24 
    // ...
}
```

Expected Results:

It is expected that all trades exiting due to stop_loss should reflect the stop-loss value actually returned by the custom_stoploss function (e.g., -35% or a dynamically calculated tighter value based on ATR and caps). If the custom_stoploss function were to fallback to self.stoploss (-0.99), it should have logged one of its defined internal warning/error conditions, which are not present in the logs for these 18 trades. The discrepancy lies in these 18 trades closing at a ~-99% loss without a corresponding log message from custom_stoploss explaining this fallback.

**Comment by xmatthias** [comment]
well to look at results properly, we'll need the backtesting .zip and .meta.json files. That'll contain all necessary information.
just the file alone makes it cumbersome to work with - as freqtrade won't easily load the result automatically.

so without looking at the results properly - i'd say that it's happening due to the pretty high leverage (at least based on the trades your logs show).
This can cause the stoploss to trigger at the entering candle (trade_duration = 0) - at which point (based on how you impelement stoploss) - your `custom_stoploss()` function was never called for this function.

Freqtrade does / would [support this](https://www.freqtrade.io/en/stable/strategy-callbacks/#adjust-stoploss-after-position-adjustments) - but not in the way you implement your custom stoploss function (you're using the old function header, which, for compatibility reasons, is not being called before the first "did it stop" evaluation).

So as far as i can tell so far (and what you can tell from the labels i applied) - i currently don't see a problem with freqtrade - as it's doing what it should - it acts in "compatibility" mode for `custom_stoploss()` - not calling custom stop right after the order filled, but evaluating the existing stoploss first.

**Comment by Sakuleta** [comment]
[backtest-result-2025-05-26_22-12-33.json](https://github.com/user-attachments/files/20451518/backtest-result-2025-05-26_22-12-33.json)
[backtest-result-2025-05-26_22-12-33_config.json](https://github.com/user-attachments/files/20451519/backtest-result-2025-05-26_22-12-33_config.json)

[backtest-result-2025-05-26_22-12-33.meta.json](https://github.com/user-attachments/files/20451527/backtest-result-2025-05-26_22-12-33.meta.json)
[backtest-result-2025-05-26_22-12-33.zip](https://github.com/user-attachments/files/20451528/backtest-result-2025-05-26_22-12-33.zip)

**Comment by xmatthias** [comment]
![Image](https://github.com/user-attachments/assets/32c7d2e2-4665-4a41-9739-ac9359074f1f)

Essentially - this confirms exactly what i said above :)

**Comment by Sakuleta** [comment]
Thank you.
